proposals/NNNN-unlock-existential-types-for-all-protocols.md

```diff
@@ -0,0 +1,271 @@
+
+
+### Proposed Solution
+
+The constraint prohibiting the use of some protocols as types will be lifted. Consequently, boilerplate code in many projects — especially libraries and frameworks — will be significantly reduced.
+
+
+### Detailed Design 
+
+The compiler will no longer differentiate between protocols that don’t have `Self` or associated type requirements and those that do. However, some restrictions will apply to the use of requirements referencing associated types as seen in the below examples.
+
+#### Examples:
+
+##### Protocol with `Self` and Associated Type Requirements 
+
+```swift
+protocol A {
+  associatedtype A
+    
+  var a: A { get }
+}
+
+struct S : A { 
+  let a: Int = 5
+}
+
+let myA: A = S() ✅ 
+
+
+let a: Any = myA.a ❌
+// We don’t know what associated type 
+// 'A' is on the existential type of 'A'.
+
+
+extension A {
+  var opaqueA: some Any {
+    a 
+  }
+  // Note that it references 
+  // the associated type 'A'.
+}
+
+let opaqueA: some Any = myA.opaqueA ❌
+// Opaque result type don't 
+// type-erase; they just conceal
+// the underlying value from the
+// user. As a result, the above 
+// is not allowed.
+```
+
+##### Protocol with Known Associated Types
+
+```swift
+protocol A {
+  associatedtype A
+
+  var a: A { get }
+}
+
+protocol RefinedA : A
+  where A == Int {}
+
+struct S : RefinedA { 
+  let a: Int = 5
+}
+
+let myRefinedA: RefinedA = S() ✅
+
+let intA: Int = myRefinedA.a ✅
+// Here we know that the associated
+// type 'A' of 'RefinedA' is 'Int'.
+```
+
+##### Protocol Composition 
+
+```swift
+protocol A {
+  associatedtype A
+
+  var a: A { get }
+}
+
+protocol RefinedA : A
+  where A == Int {}
+
+protocol B {
+  associatedtype B
+
+  var b: B { get }
+}
+
+struct S : RefinedA & B { 
+  let a: Int = 5
+  let b: Int = 5
+}
+
+
+let myRefinedAB: RefinedA & B = S() ✅
+
+
+let a: Int = myRefinedAB.a ✅
+
+let b: some Any = myRefinedAB.b ❌
+// We don’t know what type 'B' is
+// on the type-erased value 'myRefinedAB'.
+```
+
+> **Note on conflicting associated types**: Consider the following sample:
+
+  ```swift
+  ...
+
+  protocol AFixedToInt : A 
+    where A == Int {}
+
+  protocol AFixedToBool : A 
+    where A == Bool {}
+
+  let composition: AFixedToInt & AFixedToBool ⚠️
+  // The associated type 'A' has conflicting 
+  // implementations as both 'Int' and 'Bool'
+  // in this composition type. 
+
+  ```
+  > The above composition type emits a warning, as providing a value bound to the associated type '`A`' is not possible. It was also disussed to outright emit an error in this case; however, due to source compatibility constraints, the fact that the two protocols' associated types '`A`' are formally distinct and potential for protocol member disambiguation in the future, warnings were chosen.
```

Comment:
An analogy to `Never` might be good here—the type can still exist, even though there are no possible values of the type. Also, in this case, the associated types aren't really formally distinct because they come from the same base protocol. Another example where protocol composition introduces associated types with the same name but different origins might be:

```
protocol A1 { associatedtype A }
protocol A2 { associatedtype A }

protocol A1Int: A1 where A == Int {}
protocol A2Bool: A2 where A == Bool {}

let compo: A1Int & A2Bool
```

Reply:
Sorry for taking so long to reply. You're right, I agree that `Never` would really help explain this behavior. Also, we'll correct the example in the new proposal, thank you!